src/metabase/sync_database/cached_values.clj

```diff
@@ -0,0 +1,136 @@
+            [schema.core :as schema]
+            [toucan.db :as db]
+            [metabase.db.metadata-queries :as queries]
+            [metabase.sync-database.classify :as classify]
+            [metabase.sync-database.infer-special-type :as infer-special-type]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;;; Drivers use these to get field values
+
+(def ^:private ^:const ^Integer field-values-entry-max-length
+  "The maximum character length for a stored `FieldValues` entry."
+  100)
+
+(def ^:private ^:const ^Integer field-values-total-max-length
+  "Maximum total length for a FieldValues entry (combined length of all values for the field)."
+  (* classify/low-cardinality-threshold field-values-entry-max-length))
+
+(defn field-values-below-low-cardinality-threshold? [non-nil-values]
+  (and (<= (count non-nil-values) classify/low-cardinality-threshold)
+      ;; very simple check to see if total length of field-values exceeds (total values * max per value)
+       (let [total-length (reduce + (map (comp count str) non-nil-values))]
+         (<= total-length field-values-total-max-length))))
+
+(defn test-for-cardinality?
+  "Should FIELD should be tested for cardinality?"
+  [{:keys [base_type] :as field}]
+  (or (field-values/field-should-have-field-values? field)
+      (and (not (isa? base_type :type/DateTime))
+           (not (isa? base_type :type/Collection))
+           (not (= base_type :type/*)))))
+
+;; was part of test:cardinality-extract-field-values
+(defn extract-field-values
+  "Extract field-values for FIELD.  If number of values exceeds `low-cardinality-threshold` then we return an empty set of values."
+  [{:keys [name base_type] :as field} field-stats]
+  ;; TODO: we need some way of marking a field as not allowing field-values so that we can skip this work if it's not appropriate
+  ;;       for example, :type/Category fields with more than MAX values don't need to be rescanned all the time
+  (let [non-nil-values  (when (field-values/field-should-have-field-values? field)
+                          (filter identity (queries/field-distinct-values field (inc classify/low-cardinality-threshold))))
+        ;; only return the list if we didn't exceed our MAX values and if the the total character count of our values is reasable (#2332)
+        distinct-values (when (field-values-below-low-cardinality-threshold? non-nil-values)
+                          non-nil-values)]
+    (if (seq distinct-values)
+      (assoc field-stats :values distinct-values)
+      field-stats)))
+
+(defn make-field-extractor
+  "Make a generic implementation of `analyze-table`."
+  {:style/indent 1}
+  [driver & {:keys [calculate-row-count?]
+             :or   {calculate-row-count?  true}}]
+  (fn [driver table new-field-ids]
+    {:fields    (for [id new-field-ids]
+                  (extract-field-values (field/Field id) {:id id}))}))
```

Comment:
So it looks like `calculate-row-count?` isn't even being used for anything. What's the reason for having it?

Also why do we need this `make-field-extractor` function at all? It seems like every time it more or less always produces the same function. Seems to me like it could be removed altogether and its behavior could just be rolled into `generic-analyze-table`